Introduce a CONFIGURED state after a job has been allocated

WilliamMcCumstie opened the pull request:
The new `CONFIGURED` state is used to differentiate jobs which have been allocated but not started.

This also adds the new `configured?` method to jobs, which is similar but not analogous to `allocated?`. The `allocated?` method checks if an `Allocation` exists for a job. Where the `configured?` method is only true for jobs which have an `Allocation` which aren't `RUNNING`/`CANCELLING` etc.

In practice there is only a brief window where the jobs are in this state. The jobs are started immediately after allocation and transition to `RUNNING`. Currently the daemons do not report back if they have received the job so are unlikely to block the "start" message. This may change in the future.

I have manually confirmed the new `CONFIGURED` state can be displayed by the client. This was difficult to reproduce and required code alterations to slow down the controllers execution. These changes have not been committed.

WilliamMcCumstie committed:
Flag jobs as CONFIGURING after they are allocated

WilliamMcCumstie committed:
Update the long polling to include configuring

WilliamMcCumstie committed:
Include configuring jobs in `running_tasks_for`

This is the same behaviour as before but with the new configuring?
method

WilliamMcCumstie committed:
Transition jobs to CONFIGURING in AllocationRegistry

This ensures the state remains consistent with the updating
of the registry

WilliamMcCumstie committed:
Flag jobs as RUNNING after submitting to the daemons

benarmston commented on lib/flight_scheduler/submission/job.rb:
```diff
@@ -33,10 +33,13 @@ def initialize(allocation)
       @allocation.nodes.each do |node|
         initialize_job_on(node)
       end
+      # XXX Should a job transition to RUNNING after all the daemons have been
+      #     notified? Currently the daemons do not report back "job received",
+      #     so doing so reliable is not possible
+      @job.state = 'RUNNING'
```
If the message to the daemon cannot be sent, the `initialize_job_on` method will raise an exception.  So if we get to the `@job.state = 'RUNNING'` line we can be confident that all daemon's have received the message.

Is there something else that we need to do to consider this reliable?

Obviously, there is a pre-existing issue of what to do should one of the allocated nodes be down.  But that is a separate issue than is being worked on here.

benarmston replied:
What are the implications of setting the job's state before starting the batch script?  I'm thinking in particular here if the daemons can be allocated the job, but the batch script cannot be started FSR.  What is the correct state of the job?

WilliamMcCumstie replied:
Right we can probably rely on the underlining TCP socket to ensure the message is delivered. So we can reasonable assume it was submitted correctly if everything exits correctly.

I have moved the state change to the end: 323553bf8545910e213a1d993f2ec0991f88e640

In regards to the `batch_script` failing, it should raise an error a be caught in the existing handler. Not sure if this is correct, so I have added a note about it.

benarmston commented on app.rb:
```diff
@@ -347,7 +347,12 @@ def validate!
+          #
+          # Consider refactoring to change the name of the flag. Blocking
+          # on both is still the desired behaviour
+          while resource.pending? || resource.configuring? do
```
I think that this reads absolutely fine: "while pending or configuring, sleep for a bit".  We don't need the flag to change or the comment about what this code used to look like.

WilliamMcCumstie replied:
ea5210c42d461a9959af1f722722514127b26e50

benarmston commented on lib/flight_scheduler/allocation_registry.rb:
```diff
@@ -93,6 +93,17 @@ def add(allocation)
       # Duplicate the allocation so it is safe to modify
       allocation = allocation.dup
 
+      # NOTE: Updating the state of the internal job object violates the Law of Demeter
+      #       as the registry needs to "reach through" the allocation first.
+      #
+      #       Whilst not ideal, it is important the AllocationRegistry and CONFIGURING
+      #       state stay in sync with each other.
+      #
+      #       Consider refactoring to do the following:
+      #       1. Infer the Job's state directly from the registry, or
+      #       2. Change add(allocation) to build(job, nodes). This would likely require
+      #          the use of 'Reservations' instead of 'Allocations' in the scheduler
+      allocation.job.state = 'CONFIGURING' if allocation.job.pending?
```
If that bothers you, changing the state in `BaseScheduler#allocate_job` would be an alternative.  The scheduler is the only place that should be adding allocations to the registry so they would still be kept in sync with each other. Also, having this code in the allocation registry also means that the allocation registries scope has increased from "maintaining a registry of allocations", to "maintaining a registry of allocations and adjusting jobs".

Why do you think this is better than having the base scheduler manage the job?  The scheduler already manages, the job's `reason_pending`, `start_time`.  So why not this part of the state too?

WilliamMcCumstie replied:
I've taken your recommendation and moved the state change into the `BaseScheduler`.

3c536e8086d2b46319adadbd6ff9df2335e60b31

benarmston commented on lib/flight_scheduler/allocation_registry.rb:
```diff
@@ -193,6 +204,8 @@ def max_parallel_per_node(job, node, allocations: nil, excluding_jobs: nil, rese
   end
 
+  # NOTE: This method MUST not be called alone! The AllocationRegistry must be saved in
+  #       tandem with the JobRegistry. Failure to do so can lead to an inconsistent state
```
Why would it *always* be wrong to save this registry without saving the job registry?

When a single node is deallocated from a job, the allocation registry will have changed, but not necessarily the job registry?  Am I missing something here?

WilliamMcCumstie replied:
On review I agree with your assessment that there are likely situations where a registry maybe saved individually without leading to an inconsistent state.

However in practice these situations are hard to predict. Your example where a single `node` is removed from an `allocation` is a moot point because the neither registry are saved at this point.

Due to the ad-hoc nature of when the registries are saved, it is not possible to make a guarantee about their states.

As such I have reduced the specification from `MUST NOT` to `SHOULD NOT` and ensured the advice is followed across the code base.

3a5c3c6c4c416179873ff279e04fc74f9a6617e0

benarmston commented on lib/flight_scheduler/job_registry.rb:
```diff
@@ -110,9 +110,11 @@ def tasks_for(job)
     end
   end
 
+  # NOTE: This method is a misnomer, it includes jobs which may not have started but will
+  # imminently. Consider refactoring
```
So this method is used for:

1. Cancelling array jobs.  We iterate over all running tasks and cancel the task.  This introduces a potential race condition where we attempt to cancel a task which is configuring.  We have a similar issue with non-array jobs too.
2. Generating the environment for array tasks.  In particular the `JOB_NODES`/`JOB_NUM_NODES` and `JOB_NODELIST` variables.  Looking how we're generating those, I'm not convinced that we're allocating array jobs/tasks correctly (consistently with slurm).

All of that is thinking out loud.  The uses of this method need to be reconsidered.

WilliamMcCumstie replied:
Something to consider for another time

benarmston commented on lib/flight_scheduler/job_registry.rb:
```diff
@@ -135,6 +137,9 @@ def load
 
+  # NOTE: This method MUST not be called alone! The JobRegistry must be saved in
+  #       tandem with the AllocationRegistry. Failure to ndo so can lead to an
+  #       inconsistent state
```
Same question as above.  Why is it *always* wrong to call this method by itself?  For instance, why would saving the job registry alone be wrong after a job step is started?

WilliamMcCumstie replied:
Ditto 3a5c3c6c4c416179873ff279e04fc74f9a6617e0

benarmston commented on lib/flight_scheduler/allocation_registry.rb:
```diff
@@ -93,6 +93,17 @@ def add(allocation)
+      #       Whilst not ideal, it is important the AllocationRegistry and CONFIGURING
+      #       state stay in sync with each other.
+      #
+      #       Consider refactoring to do the following:
+      #       1. Infer the Job's state directly from the registry, or
+      #       2. Change add(allocation) to build(job, nodes). This would likely require
+      #          the use of 'Reservations' instead of 'Allocations' in the scheduler
+      allocation.job.state = 'CONFIGURING' if allocation.job.pending?
```
What do you mean by inferring the job's state from the registry?  Something like this?

```
def state
  return super unless super == 'PENDING'
  allocated? ? 'CONFIGURING' : 'PENDING'
end
```

WilliamMcCumstie replied:
Something along those lines would work, however I was hesitant to break the established design pattern around the `state`.

benarmston reviewed (commented):
This looks like a good step in the right direction.

I've added a few comments / questions below.

I think you're right about the job registry and allocation registry needing to be merged into a single persistence file.  However, you state that they `MUST` be saved together.  Either this is overstating the case, or I've missed something.

There are also a couple of other locations that should be adjusted for the new state.

* `BaseScheduler#schedule_event_scripts`, the code can be simplified.
* `libexec/slack.sh` line 106(ish) doesn't include `CONFIGURING` jobs.

WilliamMcCumstie committed:
Remove long polling comment, it is fine as is

WilliamMcCumstie committed:
Reduce coupling spec for Job/Allocation registry

On review there are a couple of situations where it is fine to
save one registry instead of the other. But it still isn't a good
idea

WilliamMcCumstie committed:
Set the state to RUNNING at the end of the Submission

WilliamMcCumstie committed:
Transition the jobs to CONFIGURING in BaseScheduler

WilliamMcCumstie committed:
Remove the state check along with 'reason_pending'

The 'reason_pending' already preforms a state check

WilliamMcCumstie committed:
Trigger the insufficient script with PartitionNodeLimit

Both Resources and PartitionNodeLimit mean there are insufficient
resources to fulfill the request with a slightly different
nuance

WilliamMcCumstie committed:
Consider CONFIGURING jobs as good as RUNNING in slack

This is a very brief state which will imminently be RUNNING

As such it doesn't need a different section

benarmston commented:
This LGTM now.  Thanks @WilliamMcCumstie
